examples/vision/depth_estimation.py

```diff
@@ -0,0 +1,516 @@
+        return [self.loss_metric]
+
+    def train_step(self, batch_data):
+        input, target = batch_data
+        with tf.GradientTape() as tape:
+            pred = self.model(input, training=True)
+
+            # Edges
+            dy_true, dx_true = tf.image.image_gradients(target)
+            dy_pred, dx_pred = tf.image.image_gradients(pred)
+            weights_x = tf.exp(tf.reduce_mean(tf.abs(dx_true)))
+            weights_y = tf.exp(tf.reduce_mean(tf.abs(dy_true)))
+
+            # depth_smoothness
+            smoothness_x = dx_pred * weights_x
+            smoothness_y = dy_pred * weights_y
+
+            l_edges = tf.reduce_mean(abs(smoothness_x)) + tf.reduce_mean(
+                abs(smoothness_y)
+            )
+
+            # Structural similarity (SSIM) index
+            l_ssim = tf.reduce_mean(
+                1
+                - tf.image.ssim(
+                    target,
+                    pred,
+                    max_val=config.WIDTH,
+                    filter_size=7,
+                    k1=0.01 ** 2,
+                    k2=0.03 ** 2,
+                )
+            )
+            # Point-wise depth
+            l1_loss = tf.reduce_mean(tf.abs(target - pred))
+
+            # Weights
+            w1 = 0.85
+            w2 = 0.1
+            w3 = 1.0
+
+            loss = (w1 * l_ssim) + w2 * l1_loss + (w3 * l_edges)
+
+        gradients = tape.gradient(loss, self.model.trainable_variables)
+        self.optimizer.apply_gradients(zip(gradients, self.model.trainable_variables))
+        self.loss_metric.update_state(loss)
+        return {
+            "Loss": self.loss_metric.result(),
+        }
+
+    def call(self, x):
+        pass
+
+
+"""
+## Model training
+"""
+
+optimizer = tf.keras.optimizers.Adam(
+    learning_rate=config.LR,
+    amsgrad=True,
+)
+model = UnetModel(config.HEIGHT, config.WIDTH)
+DEM = DepthEstimationModel(model=model)
+DEM.compile(optimizer)
+
+train_loader = DataGenerator(
+    data=df[:600].reset_index(drop="true"),
+    batch_size=config.BATCH_SIZE,
+    dim=(config.HEIGHT, config.WIDTH),
+)
+DEM.fit(train_loader, epochs=config.EPOCHS)
+
+"""
+## Visualizing Model output.
+Visualizing model output over validation set.
+The first image is the RGB image, the second image is the ground truth depth-map image
+and the third one is the predicted depth-map image.
+"""
+
+test_loader = next(
+    iter(
+        DataGenerator(
+            data=df[601:].reset_index(drop="true"),
+            batch_size=6,
+            dim=(config.HEIGHT, config.WIDTH),
+        )
+    )
+)
+visualizeDepthMap(test_loader, test=True)
+
+test_loader = next(
+    iter(
+        DataGenerator(
+            data=df[701:].reset_index(drop="true"),
+            batch_size=6,
+            dim=(config.HEIGHT, config.WIDTH),
+        )
+    )
+)
+visualizeDepthMap(test_loader, test=True)
+
+"""
+## Scopes of Improvement
+
+1. From the research papers that I read while I was working on this topic the encode part
```

Comment:
I would rewrite these two sections to be a bit less conversational and more information dense. Prefer second person (E.g. You could improve this model by...)